fish-rust/src/topic_monitor.rs

```diff
@@ -0,0 +1,640 @@
+
+fn topic_to_bit(t: topic_t) -> topic_bitmask_t {
+    1 << t.repr
+}
+
+// Some stuff that needs to be protected by the same lock.
+#[derive(Default)]
+struct data_t {
+    /// The current values.
+    current: generation_list_t,
+
+    /// A flag indicating that there is a current reader.
+    /// The 'reader' is responsible for calling sema_.wait().
+    has_reader: bool,
+}
+
+/// Sentinel status value indicating that a thread is waiting and needs a wakeup.
+/// Note it is an error for this bit to be set and also any topic bit.
+const STATUS_NEEDS_WAKEUP: u8 = 128;
+type status_bits_t = u8;
+
+#[derive(Default)]
+pub struct topic_monitor_t {
+    data_: Mutex<data_t>,
+
+    /// Condition variable for broadcasting notifications.
+    /// This is associated with data_'s mutex.
+    data_notifier_: Condvar,
+
+    /// A status value which describes our current state, managed via atomics.
+    /// Three possibilities:
+    ///    0:   no changed topics, no thread is waiting.
+    ///    128: no changed topics, some thread is waiting and needs wakeup.
+    ///    anything else: some changed topic, no thread is waiting.
+    ///  Note that if the msb is set (status == 128) no other bit may be set.
+    status_: AtomicU8,
+
+    /// Binary semaphore used to communicate changes.
+    /// If status_ is STATUS_NEEDS_WAKEUP, then a thread has commited to call wait() on our sema and
+    /// this must be balanced by the next call to post(). Note only one thread may wait at a time.
+    sema_: binary_semaphore_t,
+}
+
+/// The principal topic monitor.
+/// Do not attempt to move this into a lazy_static, it must be accessed from a signal handler.
+static mut s_principal: *const topic_monitor_t = std::ptr::null();
+
+/// Create a new topic monitor. Exposed for the FFI.
+pub fn new_topic_monitor() -> Box<topic_monitor_t> {
+    Box::new(topic_monitor_t::default())
+}
+
+impl topic_monitor_t {
+    /// Initialize the principal monitor, and return it.
+    /// This should be called only on the main thread.
+    pub fn initialize() -> &'static Self {
+        unsafe {
+            if s_principal.is_null() {
+                // We simply leak.
+                s_principal = Box::into_raw(new_topic_monitor());
+            }
+            &*s_principal
+        }
+    }
+
+    pub fn post(&self, topic: topic_t) {
+        // Beware, we may be in a signal handler!
+        // Atomically update the pending topics.
+        let topicbit = topic_to_bit(topic);
+        const relaxed: Ordering = Ordering::Relaxed;
+
+        // CAS in our bit, capturing the old status value.
+        let mut oldstatus: status_bits_t = 0;
+        let mut cas_success = false;
+        while !cas_success {
+            oldstatus = self.status_.load(relaxed);
+            // Clear wakeup bit and set our topic bit.
+            let mut newstatus = oldstatus;
+            newstatus &= !STATUS_NEEDS_WAKEUP; // note: bitwise not
+            newstatus |= topicbit;
```

Comment:
consider `core::ops::Not` et al... explicit method calls can remove the syntactic confusion of `!` meaning logical (not bitwise) negation.
```suggestion
            // Clear wakeup bit and set our topic bit.
            let newstatus = oldstatus.bitand(STATUS_NEEDS_WAKEUP.not()).bitor(topicbit);
```

There're of course multiple ways to write this — but a single line plays well with its comment `// Clear wakeup bit and set our topic bit.`